mithril-client/src/cardano_database_client/download_unpack.rs

```diff
@@ -0,0 +1,1046 @@
+        &self,
+        file_downloader: Arc<dyn FileDownloader>,
+        file_downloader_uris_chunk: Vec<(ImmutableFileNumber, FileDownloaderUri)>,
+        compression_algorithm: &CompressionAlgorithm,
+        immutable_files_target_dir: &Path,
+        download_id: &str,
+    ) -> MithrilResult<BTreeSet<ImmutableFileNumber>> {
+        let mut immutable_file_numbers_downloaded = BTreeSet::new();
+        let mut join_set: JoinSet<MithrilResult<ImmutableFileNumber>> = JoinSet::new();
+        for (immutable_file_number, file_downloader_uri) in file_downloader_uris_chunk.into_iter() {
+            let file_downloader_uri_clone = file_downloader_uri.to_owned();
+            let compression_algorithm_clone = compression_algorithm.to_owned();
+            let immutable_files_target_dir_clone = immutable_files_target_dir.to_owned();
+            let file_downloader_clone = file_downloader.clone();
+            let feedback_receiver_clone = self.feedback_sender.clone();
+            let logger_clone = self.logger.clone();
+            let download_id_clone = download_id.to_string();
+            join_set.spawn(async move {
+                    feedback_receiver_clone
+                        .send_event(MithrilEvent::CardanoDatabase(
+                            MithrilEventCardanoDatabase::ImmutableDownloadStarted { immutable_file_number, download_id: download_id_clone.clone()}))
+                        .await;
+                    let downloaded = file_downloader_clone
+                        .download_unpack(
+                            &file_downloader_uri_clone,
+                            &immutable_files_target_dir_clone,
+                            Some(compression_algorithm_clone),
+                            DownloadEvent::Immutable{immutable_file_number, download_id: download_id_clone.clone()},
+                        )
+                        .await;
+                    match downloaded {
+                        Ok(_) => {
+                            feedback_receiver_clone
+                                .send_event(MithrilEvent::CardanoDatabase(
+                                    MithrilEventCardanoDatabase::ImmutableDownloadCompleted { immutable_file_number, download_id: download_id_clone }))
+                                .await;
+
+                            Ok(immutable_file_number)
+                        }
+                        Err(e) => {
+                            slog::error!(
+                                logger_clone,
+                                "Failed downloading and unpacking immutable file {immutable_file_number} for location {file_downloader_uri:?}"; "error" => e.to_string()
+                            );
+                            Err(e.context(format!("Failed downloading and unpacking immutable file {immutable_file_number} for location {file_downloader_uri:?}")))
+                        }
+                    }
+                });
```

Comment:
maybe extracting this to a method would make the code more readable ? Something like `spawn_immutable_download` ?

Reply:
Done